### DISCUSSION TITLE
Using BigInteger in a function?

### DISCUSSION THREAD

**Discussion by theuserbl** [PowerShell/PowerShell#17809] [General]
How to use BigInteger in a function?

You know, that Python can use BigInteger for exponential numbers.
For example with
`848453465473638756857857879848948948936546354678948935735745748489435**1894`

But how is that possible with PowerShell?

The normal pow function can not handle big integer:
```
[bigint] $a = [bigint]::parse("848453465473638756857857879848948948936546354678948935735745748489435")
[bigint] $b = [bigint]::parse("1894")

[math]::pow($a,$b)
```

The result os it is ∞

But I can calculate it by the PowerShell with

```
[bigint] $a = [bigint]::parse("848453465473638756857857879848948948936546354678948935735745748489435")
[bigint] $b = [bigint]::parse("1894")

[bigint] $i = 0
[bigint] $number=1
for ($i=1; $i -le $b; $i+=1) {
  $number*=$a
}

Write-Output ($number)
```

That works fine, but I want to use it more often and so I want a function like the \[math\]::pow() one.

So I exclude it in a function

```
function BigPow([bigint] $basis, [bigint] $exponent) {
  [bigint] $i = 0
  [bigint] $number=1
  for ($i=1; $i -le $exponent; $i+=1) {
    $number*=$basis
  }
  return [bigint] $number
}

[bigint] $a = [bigint]::parse("848453465473638756857857879848948948936546354678948935735745748489435")
[bigint] $b = [bigint]::parse("1894")

Write-Output ( BigPow($a, $b) )
```

But it outputs an error:
```
Line |
  13 |  Write-Output ( BigPow($a, $b) )
     |                       ~~~~~~~~
     | Cannot process argument transformation on parameter 'basis'. Cannot convert the "System.Object[]"
     | value of type "System.Object[]" to type "System.Numerics.BigInteger".
```

I have then changed the function to

```
function BigPow {
  Param (
    [bigint] $basis,
    [bigint] $exponent
  )

  [bigint] $i = 0
  [bigint] $number=1
  for ($i=1; $i -le $exponent; $i+=1) {
    $number*=$basis
  }
  return [bigint] $number
}

[bigint] $a = [bigint]::parse("848453465473638756857857879848948948936546354678948935735745748489435")
[bigint] $b = [bigint]::parse("1894")

Write-Output ( BigPow($a, $b) )
```

But the result is the same.

What can I more do, then explicitly saying the function, that it is an BigInteger?

**Comment by Northman-de** [comment]
This works: `Write-Output ( BigPow $a  $b )`

You call BigPow with an array of `$a, $b`. This will go to your function variable $basis and can not be converted. Every comma in PowerShell creates an array.
`$array | format-table Name,Length` is equal to 
```
$attr = @("Name","Length")
$array | ft $attr
```

**Comment by vexx32** [comment]
Also, BigInteger has its own Pow function as it's a specialised numeric type; `[bigint]::Pow($a, $power)` (see the [BigInteger.Pow docs](https://docs.microsoft.com/en-us/dotnet/api/system.numerics.biginteger.pow?view=net-6.0) for details)
